File: Day 3/3-12-2024_Part2.py
import re
def calculate_mul_sum_with_conditions(file_path):
    # Regular expressions for mul, do, and don't instructions
    mul_pattern = r"mul\((\d+),(\d+)\)"
    do_pattern = r"do\(\)"
    dont_pattern = r"don't\(\)"
    
    total_sum = 0
    mul_enabled = True  # By default, mul instructions are enabled

    try:
        with open(file_path, 'r') as f:
            data = f.read()
        
        # Combine all patterns to process in sequence
        combined_pattern = f"{do_pattern}|{dont_pattern}|{mul_pattern}"
        matches = re.finditer(combined_pattern, data)
        
        for match in matches:
            if match.group(0) == "do()":
                mul_enabled = True  # Enable future mul instructions
            elif match.group(0) == "don't()":
                mul_enabled = False  # Disable future mul instructions
            elif match.group(1) and match.group(2) and mul_enabled:
                # Process valid mul instructions if enabled
                num1, num2 = int(match.group(1)), int(match.group(2))
                total_sum += num1 * num2

    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
    except Exception as e:
        print(f"An error occurred: {e}")
    
    return total_sum

file_path = "/home/dell/CODING/Advent of Code 2024/Day 3/3-12-2024.txt"
result = calculate_mul_sum_with_conditions(file_path)
print(f"The sum of all enabled mul instructions is: {result}")

File: Day 3/3-12-2024_Part2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>

using namespace std;

int calculateMulSumWithConditions(const string& filePath) {
    // Regular expressions for mul, do, and don't instructions
    regex mulPattern(R"(mul\((\d+),(\d+)\))");
    regex doPattern(R"(do\(\))");
    regex dontPattern(R"(don't\(\))");
    
    int totalSum = 0;
    bool mulEnabled = true; // By default, mul instructions are enabled

    try {
        ifstream inputFile(filePath);
        if (!inputFile.is_open()) {
            throw ios_base::failure("File not found");
        }

        // Read the file content
        string data((istreambuf_iterator<char>(inputFile)),
                    istreambuf_iterator<char>());

        // Combine all patterns to process in sequence
        regex combinedPattern(R"((do\(\))|(don't\(\))|(mul\((\d+),(\d+)\)))");
        auto matchesBegin = sregex_iterator(data.begin(), data.end(), combinedPattern);
        auto matchesEnd = sregex_iterator();

        for (auto it = matchesBegin; it != matchesEnd; ++it) {
            smatch match = *it;

            if (match.str(1) == "do()") {
                mulEnabled = true; // Enable future mul instructions
            } else if (match.str(2) == "don't()") {
                mulEnabled = false; // Disable future mul instructions
            } else if (match.str(3).find("mul") != string::npos && mulEnabled) {
                // Process valid mul instructions if enabled
                int num1 = stoi(match.str(4));
                int num2 = stoi(match.str(5));
                totalSum += num1 * num2;
            }
        }

    } 
    catch (const exception& e) {
        cerr << "An error occurred: " << e.what() << endl;
    }

    return totalSum;
}

int main() {
    string filePath = "/home/dell/CODING/Advent of Code 2024/Day 3/3-12-2024.txt";
    int result = calculateMulSumWithConditions(filePath);
    cout << "The sum of all enabled mul instructions is: " << result << endl;
    return 0;
}